### DISCUSSION TITLE
Cookies not set for cross domain requests

### DISCUSSION THREAD

**Discussion by JanRuettinger** [fastapi/fastapi#6653] [Questions]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.

tldr; Cookies are not set on SPA frontend.

Fastapi version: 0.65.1
Safari 14.1.1

## Problem description
I have a SPA frontend at mydomain.com which communicates with a backend at api.mydomain.com. I am using the fastapi cors middleware which seems to work fine except for cookies. When I do a post request with `axios` no cookies are set in the frontend. However, it works fine on localhost with slightly different cookie settings (no secure).

EDIT:
Sometimes cookies are set for a split second. When I inspect them in the browser I can't see them but immediately after my signup call which sets the cookie is immediately followed by another call which requires cookies. Sometimes the second call goes through.

### Axis config
```
const api = Axios.create({
  baseURL: "URL_BLABLA",
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
  withCredentials: true,
})
```

### Fastpi cors config
```
app = FastAPI(docs_url=None, redoc_url=None, openapi_url=None)

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)
app.include_router(authentification.router)
app.include_router(users.router)
app.include_router(queries.router)
```

### Fastapi route
```
@router.post("/signup")
async def signup_for_access_token(
    user_data: UserCreate,
    response: Response,
):

    access_token_expires = timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES)
    access_token = "blabla"
    response.set_cookie(
        key="access_token",
        value=access_token,
        expires=int(access_token_expires.total_seconds()),
        httponly=True,
        secure=True,
        samesite=None, # bug -> doesn't work, need to set it manually (below)
    )
    # just adding samesite flag to set-cookie headers
    for idx, header in enumerate(response.raw_headers):
        if header[0].decode("utf-8") == "set-cookie":
            cookie = header[1].decode("utf-8")
            if "SameSite=None" not in cookie:
                cookie = cookie + "; SameSite=None"
                response.raw_headers[idx] = (header[0], cookie.encode())
    print(response.__dict__)
    return {"msg": "successful"}
```

output: `print(response.__dict__)
```
{'status_code': None, 'background': None, 'body': b'', 'raw_headers': [(b'set-cookie', b'access_token=blabla_accesstoken; expires=Sat, 29 May 2021 08:59:03 GMT; HttpOnly; Path=/; Secure; SameSite=None')]}
```
=> HttpOnly, Secure and SameSite are set correctly.

I went through all relevant results on google and couldn't find a solution.

**Comment by Hultner** [comment]
As for your "bug" that's because you use samesite=None (send no value) while the cookie parameters expects "none" as a string, this is not None like the void/null value in Python. 

Which domain are you using and are you using https?
I use `response.set_cookie(samesite="none", secure=True)` over https in deployed envrionments but when I run it locally over http I must use `response.set_cookie("cookie_name", "samesite"="Lax, "secure"=False)`.

**Comment by JanRuettinger** [comment]
Thanks for the hint. I have changed the set_cookie call.

#### output of print(response.dict):
```
{'status_code': None, 'background': None, 'body': b'', 'raw_headers': [(b'set-cookie', b'access_token=blabla; expires=Sat, 29 May 2021 12:16:18 GMT; HttpOnly; Path=/; SameSite=none; Secure')]}
```

So it looks like SameSite is now set correctly but the cookie is still not set in the browser.
Yes I use https in production.

EDIT:
Cookie gets send in the response now but vanishes immediately. 

Here is the response cookie:

<img width="1495" alt="Screenshot 2021-05-21 at 13 43 17" src="https://user-images.githubusercontent.com/8582703/119139144-96743600-ba3a-11eb-8ad3-c27b17a70112.png">

**Comment by Hultner** [comment]
> Thanks for the hint. I have changed the set_cookie call.
> 
> #### output of print(response.dict):
> ```
> {'status_code': None, 'background': None, 'body': b'', 'raw_headers': [(b'set-cookie', b'access_token=blabla; expires=Sat, 29 May 2021 12:16:18 GMT; HttpOnly; Path=/; SameSite=none; Secure')]}
> ```
> 
> So it looks like SameSite is now set correctly but the cookie is still not set in the browser.
> Yes I use https in production.
> 
> EDIT:
> Cookie gets send in the response now but vanishes immediately.
> 
> Here is the response cookie:
> 
> <img alt="Screenshot 2021-05-21 at 13 43 17" width="1495" src="https://user-images.githubusercontent.com/8582703/119139144-96743600-ba3a-11eb-8ad3-c27b17a70112.png">

But are you using https when testing?
Would you like to attach the full request response from your access log?

**Comment by JanRuettinger** [comment]
I found the bug. It was not related to fastapi. My secret was not always correctly encoded. I create a new secret every time I restart the application and sometimes the encoding was wrong.

But one "bug" still remains. I can't see the cookies in the dev tools. But they are being sent correctly with every request.

Can't see or access cookies.
<img width="1456" alt="Screenshot 2021-05-21 at 20 32 11" src="https://user-images.githubusercontent.com/8582703/119189971-a3ae1680-ba74-11eb-9e1b-a820cc0a3f20.png">

But cookies are sent correctly with every request.
<img width="1502" alt="Screenshot 2021-05-21 at 20 32 39" src="https://user-images.githubusercontent.com/8582703/119190024-b58fb980-ba74-11eb-923c-794f9a0f01ad.png">

Do you know why this is the case?

**Comment by Hultner** [comment]
Not sure about safari devtools, what does it show in the Cookies tab in your screenshot there?

Maybe you’ve applied some filter?
Looks at least like it’s working for the request :)

**Comment by amanpatel123** [comment]
I have the exact same problem @JanRuettinger. I could see cookie being sent with the response object, but does not show up in dev tool. Did you manage to find the issue @JanRuettinger ?

**Comment by JunsikChoi** [comment]
If you use subdomain like 'dev.example.com', add domain=".example.com" in cookie, so the subdomain can access root domain cookie.

**Comment by ChrisB1123** [comment]
@JanRuettinger  @amanpatel123 I'm having this same problem - did you find a solution?

**Comment by artemonsh** [comment]
First of all, read [the documentation for cookies](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies). If you still have some issues, then check my solution.

I have a FastAPI app running in the production mode. The solution was to set arguments in the `response.set_cookie` like this:
```python
response.set_cookie(
                    key="token",
                    value="my_value123",
                    samesite='none', # in my case and probably in yours
                    secure=True, # if using https and not http
                    expires=60 * 60 * 24, # any number in seconds
                    domain="yourdomain.com")
```

Also take a look at [the comment above](https://github.com/tiangolo/fastapi/issues/3267#issuecomment-845872511), it may be helpful.

**Comment by yashpct2807** [comment]
Hello All, I am facing some unexpected issue as below:
**Background:**
My backend FastAPI is hosted on http://fastapi.mysite.com:8000 and Frontend React App is hosted on http://web.mysite.com over AWS EKS separate pods

I have two ways to Authenticate the User to Singin to Web Application:
1. Basi Auth - Username and Password
2. Okta SAML - Only Authentication of Okta Users

My Backend API Authorize and generate the token after successful Authentication process from any of above method.

> Above setup is working as expected in my localhost

**Problem:**
I am able to create Authorization token and set the cookie to the browser using below:
          response.set_cookie(key="access_token",value=access_token,httponly=True,secure=True,expires=60,max_age=60,domain=".mysite.com")

Above code works successfully when I do normal Authentication but not when I perform Authentication using SAML.

I have already verified the **SAMLResponse** coming from Okta and there is no issue on Authentication process. React Application is not getting token from cookie to allow user to navigate to home page.

Find below snippet of Callback API:
```python
@router.post('/saml2/callback')
async def saml_login_callback(request:Request,response:Response,db: Session = Depends(get_db)):
  form_data = await request.form()
  req = await prepare_from_saml_request(request, True)
  auth = OneLogin_Saml2_Auth(req, saml_settings)
  # Process IdP response
  auth.process_response() 
  errors = auth.get_errors() 
  # This method receives an array with the errors
  if len(errors) == 0:
    # This check if the response was ok and the user data retrieved or not (user authenticated)
    if not auth.is_authenticated(): 
      return "user Not authenticated"
    else:
       attrs = auth.get_attributes()
       username = auth.get_nameid()
       user = db.query(User).filter(User.email == username).first()
        user_id =user.id
        token_type = 'web'
        access_token = create_web_access_token(username,user_id,token_type)
       
        response = RedirectResponse(settings.OKTA_URL, status_code=302) 
        response.set_cookie("ACCESS_TOKEN", value=access_token, samesite="None",path="/",secure=True,domain=".mysite.com")
        response.headers["Location"]=settings.APP_SAML_URL
        response.headers["Access-Control-Allow-Origin"]="*"
        return response
 else:
    print("Error when processing SAML Response: %s %s" % (', '.join(errors), auth.get_last_error_reason()))
    return "Error in callback"

``` 



Can anyone please help me? @JanRuettinger  @tiangolo

**Comment by Okteeicode** [comment]
If any one still facing this problem let me know i find solution for it my stack is , react, axios and fastapi

My step

export const authSend = axios.create({
    baseURL: 'http://localhost:8000/api/v1/users/',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    withCredentials: true,
  });
  
  origins = [
    'http://localhost:3000',
    'http://192.168.1.143:3000',
]


app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)
  
response.set_cookie(key="oktee_cookie", value=f"Bearer {access_token}", httponly=True,)

**Reply by Okteeicode** [reply]
I spent the whole day on it to figure it out i hope this help someone
